# Fonksiyonlarda Return

Normalde fonksiyonlar sadece print değeri ile birlikte ekrana değer yazdırabiliyor bize herhangi bir değer vermiyordu. Ancak bu değerleri projemizde başka yerlerde kullanmak isteyebiliriz. Bu gibi durumlarda ***return*** komutu devreye giriyor.

Return ifadesi fonksiynun işlemi bittikten sonra çağrıldığı yere değer döndürmesi anlamına gelir.

In [1]:
def toplam(a,b,c):
    print("Toplam",a+b+c)

def ikiyleçarp(a):
    print("Sonuç:",a*2)    

In [3]:
toplam=toplam(1,2,3)
ikiyleçarp(toplam)

Toplam 6


TypeError: unsupported operand type(s) for *: 'NoneType' and 'int'

Burada hata almamızın sebebi fonksiyonları herhangi bir değer döndürmemesi yani return kullanmamasıdır. 

In [4]:
type(toplam)

NoneType

In [5]:
def toplama(a,b,c):
    return a+b+c # return'un kullanımı 
def ikiyle_çarp(a):
    return a*2

In [7]:
toplam = toplama(3,4,5)
print(ikiyle_çarp(toplam))

24


return ile değerleri geldiği yere döndürebilir ve o şekilde kullanabiliriz. Yukarıda toplama(3,4,5) ifadesi 12 değeri alarak kullanılır.

In [8]:
def üçleçarp(a):
    print("1.fonksiyon çalıştı")
    return a*3

In [9]:
def ikiyletopla(a):
    print("2.fonksiyon çalıştı")
    return a + 2

In [10]:
def dördeböl(a):
    print("3.fonksiyon çalıştı")
    return a / 4

In [11]:
print(dördeböl(ikiyletopla(üçleçarp(5))))

1.fonksiyon çalıştı
2.fonksiyon çalıştı
3.fonksiyon çalıştı
4.25


sırayla

print(dördeböl(ikiyletopla(üçleçarp(5))))

print(dördeböl(ikiyletopla(15))

print(dördeböl(17))

print(4.25)   

Return ifadesinden sonra fonksiyon sona erer ve ondan sonra yazıllan hiç bir işlem gerçkeleşmez.

In [14]:
def toplama(a,b,c):
    return a + b + c
    print("Toplama fonksiyonu") # Çalıştırılmadı.

print(toplama(1,2,3))    

6


Fonksiyonlarda çağırıldğı yere herhangi bir değer döndürmeyen(return kullanılmayan) Fonksiyonlara ***void*** fonksiyon denir

# Parametre Türleri

Daha önce parametre ne görmüştük ama bir parametreyi nasıl varsaıylan yapacağımız aşağıda. Tek (=) kullanılarak yapılır.


In [16]:
def selamla(isim):
    print("Merhaba",isim)

In [19]:
selamla()

TypeError: selamla() missing 1 required positional argument: 'isim'

In [23]:
def selamla(isim="isimsiz"):
    print("Merhaba",isim)

In [24]:
def bilgilerigöster(ad = "Bilgi Yok",soyad = "Bilgi Yok",numara =  "Bilgi Yok"):
    print("Ad:",ad,"Soyad:",soyad,"Numara:",numara)

In [25]:
bilgilerigöster()

Ad: Bilgi Yok Soyad: Bilgi Yok Numara: Bilgi Yok


In [26]:
bilgilerigöster("Mustafa Murat","Coşkun") # ad ve soyad değerini verdik ancak numara parametresi varsayılan değer oldu. 

Ad: Mustafa Murat Soyad: Coşkun Numara: Bilgi Yok


Böyle bir durumla karşılaşmamak için değerleri sıralı vermemiz gerekiyor ya da hangi değeri veridğimizi belirterek vermemiz gerekiyor.

In [27]:
bilgilerigöster(numara = "123456") # numara parametresini özel olarak belirtiyoruz.

Ad: Bilgi Yok Soyad: Bilgi Yok Numara: 123456


In [28]:
bilgilerigöster(ad = "Mustafa Murat",numara = "123456")

Ad: Mustafa Murat Soyad: Bilgi Yok Numara: 123456


## Esnek sayıda değerler

Normalde bir fonksiyonda önceden kaç tane değer gireceğimizi belirtememiz gerekiyor ancak bunu esnek yapabilmek için parametrenin başına *(yıldız) işareti koyarak yapabliriz.

In [29]:
def toplama(*sayılar): # Artık parametreler değişkenini bir demet gibi kullanabilirim.
    toplam =  0
    print("sayılar:",sayılar)
    for i in sayılar:
        toplam += i
    return toplam


In [30]:
print(toplama(1,2,3,4,5,6,7,8,9))

Parametreler: (1, 2, 3, 4, 5, 6, 7, 8, 9)
45


bu şekilde istediğimiz kadar değer girebiliriz.